metabase/metabase
# Move DB model persistence control to database form

Part of #22538

In order to cache model query results, people need to enable model persistence for a specific database. After that, Metabase will cache every model based on that DB data. We used to display the "Enable/disable persistence" button in the database page sidebar, together with "rescan field values", "delete database" buttons, etc.

This PR moves this control to the database form itself. Now it's going to be shown in the "Advanced options" section right under the `auto_run_queries` setting.

**Permissions hint**

Metabase keeps the model cache in a separate schema in the user's DB. This schema can be created automatically and we use a combination of the `site-uuid` setting and a database ID to generate a unique schema name. For this to work, people either need to grant Metabase permission to create schemas in their DBs or grant read-write permission to this specific schema.

This PR makes the FE display this schema name right in the "Cache models" field description. The FE utility doing that mirrors the [BE function](https://github.com/metabase/metabase/blob/master/src/metabase/driver/ddl/interface.clj#L6) implementation.

### To Verify

1. Sign in as an admin
2. Go to `/admin/databases/create` and connect a Postgres database, you can use our [sample Postgres Docker image](https://github.com/metabase/metabase-qa#postgresql-12)
3. Go to `/admin/settings/caching` and _disable_ model persistence
4. Go to `/admin/databases`, open your Postgres DB page, and click "Advanced options"
5. Ensure you don't see the "Cache models" field as it's disabled globally at the moment
6. Go to `/admin/settings/caching` and _enable_ model persistence
7. Go to `/admin/databases`, open your Postgres DB page, and click "Advanced options"
8. Ensure you can see the "Cache models" field
9. Enable persistence and click "Save changes"
10. Ensure the FE has sent `POST /api/database/:dbId/persist` request
11. Refresh the page, and ensure the toggle is enabled
12. Disable persistence and click "Save changes"
13. Ensure the FE has sent `POST /api/database/:dbId/unpersist` request
14. Refresh the page, and ensure the toggle is disabled
15. Open any other DB page (e.g. Sample Dataset), and click "Advanced options"
16. Ensure you don't see the "Cache models" field as it's not supported by other engines than Postgres
17. Go to Go to `/admin/databases/create`, select "Postgres" engine, and click "Advanced options"
18. Ensure you don't see the "Cache models" field as we're only going to show it for already connected DBs

### Demo

https://user-images.githubusercontent.com/17258145/170097717-b62ebd78-95ba-407a-b8c7-0508e69aea2e.mp4
